File: node/src/__tests__/client.test.js
// Smoke tests for @zatabox/node. Run with: npm test
const test = require('node:test');
const assert = require('node:assert');
const crypto = require('node:crypto');
const { ZataboxClient, Client, ZataboxError, VERSION } = require('../index');

function jsonRes(status, obj, headers = {}) {
  const h = new Map(Object.entries({ 'content-type': 'application/json', ...headers }));
  return Promise.resolve({
    ok: status < 400, status, statusText: '',
    headers: h, text: () => Promise.resolve(JSON.stringify(obj)),
  });
}

// Every namespace the publicly-documented SDK exposes.
const PUBLIC_NS = ['auth', 'events', 'organizer', 'eventCustomization', 'tickets',
  'orders', 'payments', 'checkin', 'community', 'growth', 'users', 'integrations', 'webhooks'];
// Internal/undocumented surfaces that must NEVER be exposed in this public SDK.
const FORBIDDEN_NS = ['admin', 'media', 'whiteLabel', 'wallets', 'scannerTokens',
  'support', 'util', 'track', 'site', 'savedSearches', 'publicEvents', 'scan', 'search', 'dataExport'];

test('constructor requires a credential', () => {
  assert.throws(() => new ZataboxClient({}), /apiKey.*bearerToken/);
});

test('base URL routing by key prefix', () => {
  assert.strictEqual(new Client({ apiKey: 'vt_test_x' }).baseUrl, 'https://sandbox.zatabox.com');
  assert.strictEqual(new Client({ apiKey: 'vt_live_x' }).baseUrl, 'https://api.zatabox.com');
  assert.strictEqual(new Client({ apiKey: 'sk_test_x' }).baseUrl, 'https://sandbox.zatabox.com');
  assert.strictEqual(new Client({ apiKey: 'vt_test_x', baseUrl: 'http://localhost:4000' }).baseUrl, 'http://localhost:4000');
});

test('exactly the documented namespaces are present', () => {
  const c = new Client({ apiKey: 'vt_live_x' });
  for (const ns of PUBLIC_NS) {
    assert.strictEqual(typeof c[ns], 'object', `missing namespace ${ns}`);
  }
  assert.strictEqual(typeof c.events.list, 'function');
  assert.strictEqual(typeof c.webhooks.verify, 'function');
  assert.strictEqual(typeof c.checkin.liveUrl, 'function');
});

test('undocumented / internal namespaces are NOT exposed', () => {
  // Platform admin, white-label, media upload, wallets, scanner tokens, etc. must
  // never ship in a public integrator SDK.
  const c = new Client({ apiKey: 'vt_live_x' });
  for (const ns of FORBIDDEN_NS) {
    assert.strictEqual(c[ns], undefined, `forbidden namespace exposed: ${ns}`);
  }
});

test('GET unwraps the data envelope and builds query', async () => {
  let seen;
  const c = new Client({ apiKey: 'vt_live_x', fetch: (url, init) => { seen = { url, init }; return jsonRes(200, { success: true, data: { items: [1] } }); } });
  const data = await c.events.list({ limit: 20, category: 'music' });
  assert.deepStrictEqual(data, { items: [1] });
  assert.match(seen.url, /\/api\/v1\/events\?limit=20&category=music$/);
  assert.strictEqual(seen.init.headers.Authorization, 'Bearer vt_live_x');
  assert.strictEqual(seen.init.method, 'GET');
});

test('writes send JSON body + auto Idempotency-Key', async () => {
  let seen;
  const c = new Client({ apiKey: 'vt_live_x', fetch: (url, init) => { seen = init; return jsonRes(201, { success: true, data: { id: 'o1' } }); } });
  await c.orders.create({ items: [] });
  assert.strictEqual(seen.headers['Content-Type'], 'application/json');
  assert.ok(seen.headers['Idempotency-Key'], 'auto idempotency key');
  assert.strictEqual(seen.body, JSON.stringify({ items: [] }));
});

test('explicit idempotency key is honoured', async () => {
  let seen;
  const c = new Client({ apiKey: 'vt_live_x', fetch: (url, init) => { seen = init; return jsonRes(201, { success: true, data: {} }); } });
  await c.orders.create({ items: [] }, { idempotencyKey: 'fixed-key' });
  assert.strictEqual(seen.headers['Idempotency-Key'], 'fixed-key');
});

test('path params are encoded', async () => {
  let seen;
  const c = new Client({ apiKey: 'vt_live_x', fetch: (url) => { seen = url; return jsonRes(200, { success: true, data: {} }); } });
  await c.events.get('a/b c');
  assert.match(seen, /\/api\/v1\/events\/a%2Fb%20c$/);
});

test('errors become ZataboxError with code/status/requestId', async () => {
  const c = new Client({ apiKey: 'vt_live_x', maxRetries: 0, fetch: () => jsonRes(404, { success: false, error: { code: 'ORDER_NOT_FOUND', message: 'nope', details: { id: 'x' } }, meta: { request_id: 'req_1' } }) });
  await assert.rejects(() => c.orders.get('x'), (e) => {
    assert.ok(e instanceof ZataboxError);
    assert.strictEqual(e.code, 'ORDER_NOT_FOUND');
    assert.strictEqual(e.status, 404);
    assert.strictEqual(e.requestId, 'req_1');
    assert.deepStrictEqual(e.details, { id: 'x' });
    return true;
  });
});

test('429 surfaces RATE_LIMITED with retryAfter', async () => {
  const c = new Client({ apiKey: 'vt_live_x', maxRetries: 0, fetch: () => jsonRes(429, { success: false, error: {} }, { 'retry-after': '30' }) });
  await assert.rejects(() => c.events.list(), (e) => {
    assert.strictEqual(e.code, 'RATE_LIMITED');
    assert.strictEqual(e.details.retryAfter, 30);
    return true;
  });
});

test('5xx retries then succeeds', async () => {
  let n = 0;
  const c = new Client({ apiKey: 'vt_live_x', maxRetries: 2, fetch: () => { n++; return n < 2 ? jsonRes(503, { success: false, error: { code: 'X' } }) : jsonRes(200, { success: true, data: { ok: true } }); } });
  assert.deepStrictEqual(await c.events.list(), { ok: true });
  assert.strictEqual(n, 2);
});

test('paginate follows both cursor shapes', async () => {
  const c = new Client({ apiKey: 'vt_live_x', fetch: (url) => {
    if (/cursor=c2/.test(url)) return jsonRes(200, { success: true, data: { items: [2], nextCursor: null } });
    return jsonRes(200, { success: true, data: { items: [1], pagination: { cursor: 'c2' } } });
  } });
  const ids = [];
  for await (const p of c.paginate(c.events.list, { limit: 10 })) ids.push(...p.items);
  assert.deepStrictEqual(ids, [1, 2]);
});

test('SSE helper returns a fully-qualified URL', () => {
  const c = new Client({ apiKey: 'vt_live_x' });
  assert.strictEqual(c.checkin.liveUrl('42'), 'https://api.zatabox.com/api/v1/checkin/event/42/live');
});

test('webhook verify accepts a valid signature and rejects a bad one', () => {
  const c = new Client({ apiKey: 'vt_live_x' });
  const secret = 'whsec_test';
  const t = Math.floor(Date.now() / 1000);
  const raw = JSON.stringify({ type: 'order.paid', id: 'evt_1' });
  const sig = crypto.createHmac('sha256', secret).update(`${t}.${raw}`).digest('hex');
  const ev = c.webhooks.verify(raw, `t=${t},v1=${sig}`, secret);
  assert.strictEqual(ev.type, 'order.paid');
  assert.throws(() => c.webhooks.verify(raw, `t=${t},v1=00`, secret), (e) => e.code === 'INVALID_SIGNATURE');
  assert.throws(() => c.webhooks.verify(raw, '', secret), (e) => e.code === 'MISSING_SIGNATURE');
});

test('setBearerToken swaps the credential', async () => {
  let auth;
  const c = new Client({ bearerToken: 'jwt1', fetch: (url, init) => { auth = init.headers.Authorization; return jsonRes(200, { success: true, data: {} }); } });
  c.setBearerToken('jwt2');
  await c.users.me();
  assert.strictEqual(auth, 'Bearer jwt2');
});

test('VERSION is exported', () => {
  assert.match(VERSION, /^\d+\.\d+\.\d+/);
});

File: go/client_test.go
package zatabox

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func newTestClient(t *testing.T, handler http.HandlerFunc) (*Client, *httptest.Server) {
	t.Helper()
	srv := httptest.NewServer(handler)
	c := New("vt_live_x", WithBaseURL(srv.URL))
	return c, srv
}

func writeEnvelope(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	body := map[string]interface{}{"success": status < 400, "data": data, "meta": map[string]string{"request_id": "req_1"}}
	_ = json.NewEncoder(w).Encode(body)
}

func TestBaseURLRouting(t *testing.T) {
	if got := New("vt_test_x").baseURL; got != defaultSandbox {
		t.Fatalf("test key -> %s", got)
	}
	if got := New("vt_live_x").baseURL; got != defaultLive {
		t.Fatalf("live key -> %s", got)
	}
	if got := New("vt_test_x", WithBaseURL("http://localhost:4000")).baseURL; got != "http://localhost:4000" {
		t.Fatalf("override -> %s", got)
	}
}

func TestGetUnwrapsAndQuery(t *testing.T) {
	var gotPath, gotAuth, gotMethod string
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.RequestURI()
		gotAuth = r.Header.Get("Authorization")
		gotMethod = r.Method
		writeEnvelope(w, 200, map[string]interface{}{"items": []int{1}})
	})
	defer srv.Close()

	data, err := c.Events.List(context.Background(), WithQuery(map[string]interface{}{"limit": 20, "category": "music"}))
	if err != nil {
		t.Fatal(err)
	}
	var out struct {
		Items []int `json:"items"`
	}
	_ = json.Unmarshal(data, &out)
	if len(out.Items) != 1 || out.Items[0] != 1 {
		t.Fatalf("unwrap failed: %s", data)
	}
	if !strings.HasPrefix(gotPath, "/api/v1/events?") || !strings.Contains(gotPath, "limit=20") {
		t.Fatalf("path %s", gotPath)
	}
	if gotAuth != "Bearer vt_live_x" {
		t.Fatalf("auth %s", gotAuth)
	}
	if gotMethod != "GET" {
		t.Fatalf("method %s", gotMethod)
	}
}

func TestWriteBodyAndIdempotency(t *testing.T) {
	var gotIdem, gotCT string
	var gotBody map[string]interface{}
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotIdem = r.Header.Get("Idempotency-Key")
		gotCT = r.Header.Get("Content-Type")
		_ = json.NewDecoder(r.Body).Decode(&gotBody)
		writeEnvelope(w, 201, map[string]interface{}{"id": "o1"})
	})
	defer srv.Close()

	if _, err := c.Orders.Create(context.Background(), WithBody(map[string]interface{}{"items": []interface{}{}})); err != nil {
		t.Fatal(err)
	}
	if gotIdem == "" {
		t.Fatal("expected auto idempotency key")
	}
	if gotCT != "application/json" {
		t.Fatalf("content-type %s", gotCT)
	}
	if _, ok := gotBody["items"]; !ok {
		t.Fatalf("body %v", gotBody)
	}
}

func TestExplicitIdempotency(t *testing.T) {
	var gotIdem string
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotIdem = r.Header.Get("Idempotency-Key")
		writeEnvelope(w, 201, map[string]interface{}{})
	})
	defer srv.Close()
	_, _ = c.Orders.Create(context.Background(), WithBody(map[string]interface{}{}), WithIdempotencyKey("fixed"))
	if gotIdem != "fixed" {
		t.Fatalf("idem %s", gotIdem)
	}
}

func TestPathEncoding(t *testing.T) {
	var gotPath string
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.EscapedPath()
		writeEnvelope(w, 200, map[string]interface{}{})
	})
	defer srv.Close()
	_, _ = c.Events.Get(context.Background(), "a/b c")
	if !strings.Contains(gotPath, "/api/v1/events/a%2Fb%20c") {
		t.Fatalf("path %s", gotPath)
	}
}

func TestErrorMapping(t *testing.T) {
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(404)
		_, _ = w.Write([]byte(`{"success":false,"error":{"code":"ORDER_NOT_FOUND","message":"nope","details":{"id":"x"}},"meta":{"request_id":"req_9"}}`))
	})
	defer srv.Close()
	c.maxRetries = 0
	_, err := c.Orders.Get(context.Background(), "x")
	if err == nil {
		t.Fatal("expected error")
	}
	var e *Error
	if !asError(err, &e) {
		t.Fatalf("not *Error: %v", err)
	}
	if e.Code != "ORDER_NOT_FOUND" || e.Status != 404 || e.RequestID != "req_9" {
		t.Fatalf("error %+v", e)
	}
}

func TestRateLimited(t *testing.T) {
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Retry-After", "30")
		w.WriteHeader(429)
		_, _ = w.Write([]byte(`{"success":false,"error":{}}`))
	})
	defer srv.Close()
	c.maxRetries = 0
	_, err := c.Events.List(context.Background())
	var e *Error
	if !asError(err, &e) || e.Code != "RATE_LIMITED" || e.RetryAfter != "30" {
		t.Fatalf("error %v", err)
	}
}

func TestBinary(t *testing.T) {
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/pdf")
		_, _ = w.Write([]byte("%PDF"))
	})
	defer srv.Close()
	data, ct, err := c.Tickets.Pdf(context.Background(), "5")
	if err != nil {
		t.Fatal(err)
	}
	if string(data) != "%PDF" || ct != "application/pdf" {
		t.Fatalf("got %q %q", data, ct)
	}
}

func TestPaginate(t *testing.T) {
	c, srv := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if strings.Contains(r.URL.RawQuery, "cursor=c2") {
			writeEnvelope(w, 200, map[string]interface{}{"items": []int{2}, "nextCursor": nil})
		} else {
			writeEnvelope(w, 200, map[string]interface{}{"items": []int{1}, "pagination": map[string]interface{}{"cursor": "c2"}})
		}
	})
	defer srv.Close()
	var ids []int
	err := c.Paginate(context.Background(), c.Events.List, map[string]interface{}{"limit": 10}, func(page json.RawMessage) error {
		var p struct {
			Items []int `json:"items"`
		}
		_ = json.Unmarshal(page, &p)
		ids = append(ids, p.Items...)
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	if fmt.Sprint(ids) != "[1 2]" {
		t.Fatalf("ids %v", ids)
	}
}

func TestSSEURL(t *testing.T) {
	c := New("vt_live_x")
	if got := c.Checkin.LiveURL("42"); got != "https://api.zatabox.com/api/v1/checkin/event/42/live" {
		t.Fatalf("sse url %s", got)
	}
}

func TestWebhookVerify(t *testing.T) {
	c := New("vt_live_x")
	secret := "whsec"
	ts := fmt.Sprint(time.Now().Unix())
	raw := []byte(`{"type":"order.paid"}`)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(ts + "." + string(raw)))
	sig := hex.EncodeToString(mac.Sum(nil))
	ev, err := c.Webhooks.Verify(raw, "t="+ts+",v1="+sig, secret)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(ev), "order.paid") {
		t.Fatalf("event %s", ev)
	}
	if _, err := c.Webhooks.Verify(raw, "t="+ts+",v1=00", secret); err == nil {
		t.Fatal("expected bad-signature error")
	}
	if _, err := c.Webhooks.Verify(raw, "", secret); err == nil {
		t.Fatal("expected missing-signature error")
	}
}

func TestSetBearerToken(t *testing.T) {
	var gotAuth string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		writeEnvelope(w, 200, map[string]interface{}{})
	}))
	defer srv.Close()
	c := New("", WithBearerToken("jwt1"), WithBaseURL(srv.URL))
	c.SetBearerToken("jwt2")
	_, _ = c.Users.Me(context.Background())
	if gotAuth != "Bearer jwt2" {
		t.Fatalf("auth %s", gotAuth)
	}
}

func TestUpload(t *testing.T) {
	var gotName, gotContent, gotField, gotCT string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotCT = r.Header.Get("Content-Type")
		if err := r.ParseMultipartForm(1 << 20); err != nil {
			t.Errorf("parse multipart: %v", err)
		}
		f, hdr, err := r.FormFile("file")
		if err != nil {
			t.Errorf("form file: %v", err)
		} else {
			b, _ := io.ReadAll(f)
			gotName = hdr.Filename
			gotContent = string(b)
		}
		gotField = r.FormValue("caption")
		writeEnvelope(w, 200, map[string]interface{}{"ok": true})
	}))
	defer srv.Close()
	c := New("vt_live_x", WithBaseURL(srv.URL))
	_, err := c.Media.Upload(context.Background(), []byte("PNGDATA"), UploadOptions{
		Filename: "cover.png", ContentType: "image/png", Fields: map[string]string{"caption": "hi"},
	})
	if err != nil {
		t.Fatal(err)
	}
	if gotName != "cover.png" || gotContent != "PNGDATA" || gotField != "hi" {
		t.Fatalf("got name=%q content=%q field=%q", gotName, gotContent, gotField)
	}
	if !strings.HasPrefix(gotCT, "multipart/form-data; boundary=") {
		t.Fatalf("content-type %q", gotCT)
	}
}

// asError is a tiny errors.As shim kept local so the test is self-contained.
func asError(err error, target **Error) bool {
	if e, ok := err.(*Error); ok {
		*target = e
		return true
	}
	return false
}
